src/MagicWormhole/Internal/Peer.hs

```diff
@@ -0,0 +1,98 @@
+-- | The peer-to-peer aspects of the Magic Wormhole protocol.
+--
+-- Described as the "client" protocol in the Magic Wormhole documentation.
+module MagicWormhole.Internal.Peer
+  ( pakeExchange
+  , Error
+  -- * Exported for testing
+  , wormholeSpakeProtocol
+  , decodeElement
+  , encodeElement
+  ) where
+
+import Protolude hiding (phase)
+
+import Control.Monad (fail)
+import Crypto.Hash (SHA256(..))
+import qualified Crypto.Spake2 as Spake2
+import Crypto.Spake2.Group (Group(Element, arbitraryElement))
+import Crypto.Spake2.Groups (Ed25519(..))
+import Data.Aeson (FromJSON, ToJSON, (.=), object, Value(..), (.:))
+import Data.Aeson.Types (typeMismatch)
+import qualified Data.Aeson as Aeson
+import Data.ByteArray.Encoding (convertToBase, convertFromBase, Base(Base16))
+import Data.String (String)
+
+import qualified MagicWormhole.Internal.Messages as Messages
+import qualified MagicWormhole.Internal.Rendezvous as Rendezvous
+
+-- | The version of the SPAKE2 protocol used by Magic Wormhole.
+type Spake2Protocol = Spake2.Protocol Ed25519 SHA256
+
+newtype Spake2Message = Spake2Message ByteString deriving (Eq, Show)
+
+instance ToJSON Spake2Message where
+  toJSON (Spake2Message msg) = object [ "pake_v1" .= toS @ByteString @Text (convertToBase Base16 msg) ]
+
+instance FromJSON Spake2Message where
+  parseJSON (Object msg) = do
+    hexKey <- toS @Text @ByteString <$> msg .: "pake_v1"
```

Comment:
This seems to be supplying some type constraints so that the right implementation of toS is selected.  But I'm not familiar with this syntax.  Does it have a name or do you know where I can read about it?

Reply:
Correct. 

The extensions is called `TypeApplications`. The GHC manual has a [pretty comprehensive section](https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/glasgow_exts.html#visible-type-application). This [blog post](https://kseo.github.io/posts/2017-01-08-visible-type-application-ghc8.html) is a pretty good high level summary.

```
λ> :t Nothing
Nothing :: Maybe a
λ> :t Nothing @Int
Nothing @Int :: Maybe Int
```